Permit Bootstrap deployment on Ubuntu

ycointe opened the pull request:
**Component**: Salt

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: MetalK8s on Ubuntu

**Summary**: We want to deploy the Bootstrap node on Ubuntu

**Acceptance criteria**: Launch the bootstrap.sh script and deploy bootstrap node on Ubuntu


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #1728 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

alexandre-allard committed:
Move offline.sls into redhat.sls

This is to prepare handling of ubuntu repositories
configuration

Issue: GH-1728

alexandre-allard committed:
Move repositories config from defaults to jinja map

Handle the repositories configuration based on
the OS grain

Issue: GH-1728

ycointe committed:
Salt/module: Split package_manager module

We split the module package_manager.py in to part, one for CentOS and
one for Ubuntu.
To allow to these modules to be used, we changed their virtualname.

Refs: #1728

ycointe committed:
Salt/state: Detect the OS to call the appropriate module

The package manager state call the appropriate package manager module
(package_manager_yum.py or package_manager_apt.py) depending of the OS.

Ref: #1728

ycointe committed:
Salt/repo: Deploy the Ubuntu repositories

Permit to deploy the internal repositories sources on the Ubuntu machine
and to give the highest priority to these repositories to be sure that
we use these ones. The option to disable official repo doesn't exist
on Ubuntu, so we use the repo priorities to have the same required effect.

Ref: #1728

ycointe committed:
Salt/module: Add DEB and RPM Package manager modules

As we splitted the package manager in two parts for CentOS and Ubuntu,
we need to declare them in the salt_tree.

Ref: #1728

ycointe committed:
Salt/packages: Sort packages used by Ubuntu and CentOS

We sort the packages used by CentOS and Ubuntu and we change the used
package name to download it depending if the package is downloaded on
CentOS or Ubuntu.

Ref: #1728

ycointe committed:
Salt/package: Permit to install containerd on Ubuntu

We allow to install `containerd` on Ubuntu, using the CentOS requirement
only for CentOS.

Ref: #1728

ycointe committed:
Bootstrap/deploy: Permit to use apiserver on Ubuntu

Permit to use apiserver on Ubuntu, removing the use of uncreated folder.

Ref: #1728

bert-e commented:
# Hello ycointe,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

TeddyAndrieux commented on salt/_modules/metalk8s_package_manager_apt.py:
```diff
@@ -0,0 +1,142 @@
+# pylint: disable=import-error
+import apt      # type: ignore
+import apt_pkg  # type: ignore
+# pylint: enable=import-error
```
Please do like for `salt/_modules/metalk8s_cordon.py` using `HAS_LIBS` because here if you not have apt you will get errors

alexandre-allard committed:
Add --resolv-conf option to kubelet

This option is needed on system using systemd-resolved to
point to the real resolvconf file, otherwise it prevents
coredns from starting as it detects a forwading loop.

More information here:
  https://github.com/coredns/coredns/tree/master/plugin/loop

Issue: GH-1728

alexandre-allard commented on salt/_modules/metalk8s_package_manager_apt.py:
```diff
@@ -0,0 +1,142 @@
+#        if not is_installed and pkg_name != name:
+            # Any package requiring the target `name` that is not yet installed
+            # should not be installed
+#            del all_pkgs[pkg_name]
```
Please remove all the commented code

alexandre-allard commented on salt/_modules/metalk8s_package_manager_apt.py:
```diff
@@ -0,0 +1,142 @@
+#    )
+#
+#    all_pkgs.update(dependents)
+
```
Remove this, it'll be handled in another PR I think

alexandre-allard commented on salt/_modules/metalk8s_package_manager_yum.py:
```diff
@@ -11,8 +11,9 @@
 def __virtual__():
-    return __virtualname__
-
+    if __grains__['os_family'].lower() == 'redhat':
```
You don't need lower() as we already know how it's written:
```
if __grains__['os_family'] == 'RedHat':
```

alexandre-allard commented on salt/metalk8s/map.jinja:
```diff
@@ -2,13 +2,31 @@
+  'Ubuntu': {
+    'm2crypto': 'python-m2crypto',
+    'iproute': 'iproute2',
+
```
extra line here

alexandre-allard commented on salt/metalk8s/map.jinja:
```diff
@@ -24,23 +42,139 @@
+    'repositories': {
+      'metalk8s-bionic': {
+        'name': 'deb [trusted=yes] '
+              'file:///vagrant/_build/root/packages/debian/metalk8s-bionic '
```
You should not store the hardcoded path here, it must be computed in the salt formula instead.

alexandre-allard commented on salt/metalk8s/map.jinja:
```diff
@@ -24,23 +42,139 @@
+  'Ubuntu': {
+    'repositories': {
+      'metalk8s-bionic': {
+        'name': 'deb [trusted=yes] '
```
Should be a dict of options (and then rendered in the formula):
```
  'Ubuntu': {
    'repositories': {
      'metalk8s-bionic': {
        'options': {
          'trusted': 'yes',
        },
```

alexandre-allard commented on salt/metalk8s/repo/ubuntu.sls:
```diff
@@ -0,0 +1,33 @@
+
+Repositories configured:
+  test.succeed_without_changes:
+    - require:
```
Must also require the state above `- file: Add APT metalk8s repositories preferences`

alexandre-allard commented on salt/metalk8s/repo/ubuntu.sls:
```diff
@@ -0,0 +1,33 @@
+    - name: python-apt
+
+{%- for repo_name, repo_config in repo.repositories.items() %}
+Configure {{repo_name}} repository:
```
Please, for consistency, put spaces between the jinja markups as it's done everywhere else.

alexandre-allard commented on salt/metalk8s/repo/ubuntu.sls:
```diff
@@ -0,0 +1,33 @@
+{%- from "metalk8s/map.jinja" import repo with context %}
+
+{%- set repo_host = pillar.metalk8s.endpoints['repositories'].ip %}
+{%- set repo_port = pillar.metalk8s.endpoints['repositories'].ports.http %}
```
These variables are not used anywhere

alexandre-allard commented on salt/metalk8s/repo/ubuntu.sls:
```diff
@@ -0,0 +1,33 @@
+{%- for repo_name, repo_config in repo.repositories.items() %}
+Configure {{repo_name}} repository:
+  pkgrepo.managed:
+    - name : {{repo_config.name}}
```
So you only handle local repositories, it seems that deploying on bootstrap node was the scope of the linked issue, but hey it does not cost much to handle http served repositories.

ycointe committed:
Remove hardcoded paths for repo

We removed hardcoded paths used for configured repositories.

Ref: #1728

alexandre-allard committed:
Serve repositories through http

Issue: GH-1728

alexandre-allard commented on salt/metalk8s/map.jinja:
```diff
@@ -24,23 +41,139 @@
+          'trusted': 'yes',
+        },
+        'comps': [
+            'metalk8s-bionic'
```
Too much indentation

alexandre-allard commented on salt/metalk8s/map.jinja:
```diff
@@ -24,23 +41,139 @@
+        'options': {
+          'trusted': 'yes',
+        },
+         'comps': [
```
extra space before `'comps'`

alexandre-allard commented on salt/_modules/metalk8s_package_manager_apt.py:
```diff
@@ -0,0 +1,117 @@
+HAS_LIBS = True
+import apt
+import apt_pkg
+HAS_LIBS = False
```
No, no, no, no, no !
```
HAS_APT_LIBS = True
try:
    import apt
    import apt_pkg
except ImportError:
    HAS_APT_LIBS = False
```
Btw, `apt_pkg` does not seem to be used anywhere.

alexandre-allard committed:
Replace pkgrepo.managed by file.managed

We no longer use pkgrepo module as it does not update
properly the repositories definitions on Ubuntu.
It appends definitions in sources.list files instead
of replacing them.

Issue: GH-1728

TeddyAndrieux commented on salt/_modules/metalk8s_package_manager_apt.py:
```diff
@@ -0,0 +1,118 @@
+
+
+def __virtual__():
+    if __grains__['os_family'].lower() == 'debian':
```
It's better to me if you use `HAS_APT_LIBS` here

TeddyAndrieux commented on salt/metalk8s/map.jinja:
```diff
@@ -2,13 +2,30 @@
+    },
   },
-} %}
+}, grain='os', merge=defaults.get('version_defaults')) %}
```
I suggest to use `os_family` instead of `os` as we expect that most of the code will work for RedHat and CentOS

TeddyAndrieux replied:
PS: If you give no `grain` arg it's `os_family` by default https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.grains.html#salt.modules.grains.filter_by

TeddyAndrieux commented on salt/metalk8s/map.jinja:
```diff
@@ -24,23 +41,145 @@
-  'default': {}
-}, merge=defaults.get('kubernetes')) %}
+  'Ubuntu': {}
+}, grain='os', merge=defaults.get('kubernetes')) %}
```
Why this change ? We want to merge this variable also for RedHat. If you have nothing special to override here keep the old one please :)

TeddyAndrieux commented on salt/metalk8s/repo/ubuntu.sls:
```diff
@@ -0,0 +1,59 @@
+
+Install python-apt:
+  pkg.installed:
+    - name: python-apt
```
Here you expect that some repo available before our configuration with `python-apt` available. Please remove this and just assume that `python-apt` is installed for the moment (it's installed with salt, so for the bootstrap node it's ok)

TeddyAndrieux commented on salt/metalk8s/repo/ubuntu.sls:
```diff
@@ -0,0 +1,59 @@
+      - file: Add APT metalk8s repositories preferences
+{%- for repository_name in repo.repositories.keys() %}
+      - file: Configure {{ repository_name }} repository
+{%- endfor %}
```
For me should be better to use `require_in` instead of doing two time the `for` loop here (but I know it's the same for RedHat for the moment :stuck_out_tongue: )

alexandre-allard commented on salt/metalk8s/repo/centos.sls:
```diff
@@ -13,12 +13,12 @@ Install yum-plugin-versionlock:
+    {%- set repo_base_url = "file://" ~
                             salt.metalk8s.get_archives()[saltenv].path ~ "/" ~
-                            repo.relative_path %}
+                            repo.relative_path ~ "/" ~ ['os_family'].lower() %}
```
```grains['os_family'].lower()```

TeddyAndrieux commented on salt/metalk8s/map.jinja:
```diff
@@ -55,8 +194,21 @@
+      }
+    }
+  }
+}, grain='os', merge=defaults.get('kubelet')) %}
```
Please use `osfamily` (or just remove `grain='os'` as `osfamily` is the default value) everywhere or @alexandre-allard-scality  change for redhat will be broken in 2.4

Even if you (or Alexandre) are going to have conflicts in any case I think.

TeddyAndrieux commented on salt/metalk8s/repo/init.sls:
```diff
@@ -1,2 +1,2 @@
 include:
-  - .offline
+  - .{{ grains['os'] | lower }}
```
For me you should use `osfamily` here also

ycointe commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye ycointe.
